Encapsulate the calls to tbb::parallel_for

Rombur opened the pull request:
This PR encapsulates the calls to `tbb::parallel_for`. The reason is to simplify a future support of `OpenMP`

kronbichler commented:
/rebuild

masterleinad commented:
`base/parallel_transform_03.debug` is failing with:

```
738: Test command: /export/home/darndt/.linuxbrew/Cellar/cmake/3.12.3/bin/cmake "-DTRGT=parallel_transform_03.debug.diff" "-DTEST=base/parallel_transform_03.debug" "-DEXPECT=PASSED" "-DBINARY_DIR=/mnt/data/darndt/Sources/dealii-clang-6.0.0/build/tests/base" "-DGUARD_FILE=/mnt/data/darndt/Sources/dealii-clang-6.0.0/build/tests/base/parallel_transform_03.debug/interrupt_guard.cc" "-P" "/mnt/data/darndt/Sources/dealii-clang-6.0.0/build/share/deal.II/scripts/run_test.cmake"
738: Test timeout computed to be: 600
738: In file included from /mnt/data/darndt/Sources/dealii-clang-6.0.0/tests/base/parallel_transform_03.cc:23:
738: In file included from /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/lambda/lambda.hpp:14:
738: In file included from /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/lambda/core.hpp:28:
738: In file included from /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/tuple/tuple.hpp:28:
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/tuple/detail/tuple_basic.hpp:133:23: error: no type named 'head_type' in 'boost::tuples::null_type'
738:       apply<T>::type::head_type type;
738:       ~~~~~~~~~~~~~~~~^~~~~~~~~
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/lambda/detail/lambda_traits.hpp:108:32: note: in instantiation of template class 'boost::tuples::element<1, boost::tuples::tuple<tbb::blocked_range<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> > > &, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> >' requested here
738:        typename boost::tuples::element<N, T>::type
738:                                ^
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/lambda/detail/lambda_functors.hpp:68:15: note: in instantiation of template class 'boost::lambda::detail::tuple_element_as_reference<1, boost::tuples::tuple<tbb::blocked_range<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> > > &, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> >' requested here
738:       detail::tuple_element_as_reference<N, Tuple>::type type;
738:               ^
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/lambda/detail/lambda_functors.hpp:94:30: note: in instantiation of template class 'boost::lambda::detail::get_element_or_null_type<1, boost::tuples::tuple<tbb::blocked_range<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> > > &, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> >' requested here
738:     typedef typename detail::get_element_or_null_type<1, SigArgs>::type type;
738:                              ^
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/lambda/detail/lambda_functor_base.hpp:178:37: note: in instantiation of template class 'boost::lambda::placeholder<2>::sig<boost::tuples::tuple<tbb::blocked_range<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> > > &, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> >' requested here
738:   typedef typename lf_t::inherited::template sig<SigArgs>::type el_t;  
738:                                     ^
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/lambda/detail/lambda_functor_base.hpp:183:16: note: in instantiation of template class 'boost::lambda::detail::deduce_argument_types_<boost::tuples::cons<boost::lambda::lambda_functor<boost::lambda::placeholder<SECOND> >, boost::tuples::null_type>, boost::tuples::tuple<tbb::blocked_range<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> > > &, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> >' requested here
738:       typename deduce_argument_types_<typename Args::tail_type, SigArgs>::type
738:                ^
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/bundled/boost-1.62.0/include/boost/lambda/detail/lambda_functor_base.hpp:227:20: note: (skipping 13 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
738:   typedef typename deduce_argument_types_<Args, SigArgs>::type t1;
738:                    ^
738: /usr/include/tbb/parallel_for.h:88:67: note: in instantiation of member function 'tbb::interface9::internal::start_for<tbb::blocked_range<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> > >, boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::minus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::plus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::placeholder<FIRST> >, boost::lambda::lambda_functor<boost::lambda::placeholder<SECOND> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > >, boost::lambda::lambda_functor<boost::lambda::placeholder<THIRD> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > >, const tbb::auto_partitioner>::start_for' requested here
738:                 start_for& a = *new(task::allocate_root(context)) start_for(range,body,partitioner);
738:                                                                   ^
738: /usr/include/tbb/parallel_for.h:200:61: note: in instantiation of member function 'tbb::interface9::internal::start_for<tbb::blocked_range<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> > >, boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::minus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::plus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::placeholder<FIRST> >, boost::lambda::lambda_functor<boost::lambda::placeholder<SECOND> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > >, boost::lambda::lambda_functor<boost::lambda::placeholder<THIRD> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > >, const tbb::auto_partitioner>::run' requested here
738:     internal::start_for<Range,Body,const auto_partitioner>::run(range,body,partitioner);
738:                                                             ^
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/include/deal.II/base/parallel.h:160:12: note: in instantiation of function template specialization 'tbb::parallel_for<tbb::blocked_range<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> > >, boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::minus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::plus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::placeholder<FIRST> >, boost::lambda::lambda_functor<boost::lambda::placeholder<SECOND> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > >, boost::lambda::lambda_functor<boost::lambda::placeholder<THIRD> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > > >' requested here
738:       tbb::parallel_for(tbb::blocked_range<Iterator>(x_begin, x_end, grainsize),
738:            ^
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/include/deal.II/base/parallel.h:346:15: note: in instantiation of function template specialization 'dealii::parallel::internal::parallel_for<dealii::SynchronousIterators<std::tuple<double *, double *, double *, double *> >, boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::minus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::plus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::placeholder<FIRST> >, boost::lambda::lambda_functor<boost::lambda::placeholder<SECOND> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > >, boost::lambda::lambda_functor<boost::lambda::placeholder<THIRD> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > > >' requested here
738:     internal::parallel_for(SyncIterators(x_begin),
738:               ^
738: /mnt/data/darndt/Sources/dealii-clang-6.0.0/tests/base/parallel_transform_03.cc:46:13: note: in instantiation of function template specialization 'dealii::parallel::transform<double *, double *, double *, double *, const boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::minus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::lambda_functor_base<boost::lambda::arithmetic_action<boost::lambda::plus_action>, boost::tuples::tuple<boost::lambda::lambda_functor<boost::lambda::placeholder<FIRST> >, boost::lambda::lambda_functor<boost::lambda::placeholder<SECOND> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > >, boost::lambda::lambda_functor<boost::lambda::placeholder<THIRD> >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type> > > >' requested here
738:   parallel::transform(x.begin(),
738:             ^
[...]
```

Rombur commented:
Fixed. I also changed the api for `transform` it used to take a `Predicate &` which I changed to `const Predicate &`. This does not change the code because inside `transform` we required `const Predicate &`
